advanced-programming/mutexes/my_mutex.go

```diff
@@ -0,0 +1,62 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type mutex struct {
+	val int64
+}
+
+var counter int
+
+func (m *mutex) Lock() {
+	for {
+		if atomic.CompareAndSwapInt64(&m.val, 0, 1) == true {
+			return
+		}
+	}
+}
+
+func (m *mutex) Unlock() {
```

Comment:
in my solution, I put unlock into a loop 🤦 reading your solution helped me realize we should never call unlock unless we can unlock, so I added a panic if the compare and swap fails. Thanks!